### PR TITLE
feat(atomic): add & export resultContext method

### PR DESCRIPTION
**Pull request by ThibodeauJF**
https://coveord.atlassian.net/browse/KIT-1259

### CODE REVIEW

File: packages/atomic/src/components/result-template-components/result-template-decorators.tsx [hunk 1]
```diff
@@ -58,3 +58,30 @@ export function ResultContext() {
...
+ * Retrieves `Result` on a rendered `atomic-result`.
+ *
+ * This method is useful to build custom result template elements, see [Create a Result List](https://docs.coveo.com/en/atomic/latest/usage/create-a-result-list/) for more information.
+ *
+ * @param element Element on which to dispatch the event, which must be the child of a rendered "atomic-result".
+ * @returns A promise that resolves on initialization of the parent "atomic-result" element, and rejects when it's not the case.
+ */
+export const resultContext = (element: Element) =>
+  new Promise<Result>((resolve, reject) => {
+    setTimeout(() => {
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/result-template-components/result-template-decorators.tsx:76]: Something strange happening here...

Without the timeout, the result cannot be resolve, it seems that for an instant, the element exists somewhere not in the DOM, it must be linked to the way we render result templates

Comment (samisayegh) [reply packages/atomic/src/components/result-template-components/result-template-decorators.tsx:76]: Perhaps `initialize` is being called before [componentDidRender](https://stenciljs.com/docs/component-lifecycle#component-lifecycle-methods) in the case of result templates?

**Review by samisayegh — APPROVED** [review]

File: packages/atomic/src/components/result-template-components/result-template-decorators.tsx (hunk 1 above)
Comment (ThibodeauJF) [reply packages/atomic/src/components/result-template-components/result-template-decorators.tsx:76]: It's called on connectedCallback in my example, which should be "Invoked each time the custom element is appended into a document-connected element."

To be more precise, the event would not resolve (wouldn't be caught as if the parent atomic-result did not exist). document.body.contain would be false too and element.closest('atomic-result') would return nothing as well

**Review by dlutzCoveo — APPROVED** [review]

File: packages/atomic/src/components/result-template-components/result-template-decorators.tsx (hunk 1 above)
Comment (ThibodeauJF) [reply packages/atomic/src/components/result-template-components/result-template-decorators.tsx:76]:
```suggestion
    // TODO: https://coveord.atlassian.net/browse/KIT-127 make it work with connectedCallback
    setTimeout(() => {
```

File: packages/atomic/src/components/result-template-components/atomic-result-fields-list/atomic-result-fields-list.tsx [hunk 2]
```diff
@@ -11,7 +11,7 @@ import {ResultContext} from '../result-template-decorators';
 })
 export class AtomicResultFieldsList {
   @ResultContext()
-  private resizeObserver!: ResizeObserver;
+  private resizeObserver?: ResizeObserver;
```
Comment (btaillon) [inline packages/atomic/src/components/result-template-components/atomic-result-fields-list/atomic-result-fields-list.tsx:14]: I'm not sure why `@ResultContext()` is applied to `resizeObserver`, that looks like a mistake. Do you think you could remove it?

File: packages/atomic/src/components/result-template-components/result-template-decorators.tsx [hunk 3]
```diff
@@ -58,3 +58,31 @@ export function ResultContext() {
...
+ * Retrieves `Result` on a rendered `atomic-result`.
+ *
+ * This method is useful for building custom result template elements, see [Create a Result List](https://docs.coveo.com/en/atomic/latest/usage/create-a-result-list/) for more information.
+ *
+ * @param element The element that the event is dispatched to, which must be the child of a rendered "atomic-result".
+ * @returns A promise that resolves on initialization of the parent "atomic-result" element, or rejects when there is no parent "atomic-result" element.
+ */
+export const resultContext = (element: Element) =>
+  new Promise<Result>((resolve, reject) => {
+    // TODO: https://coveord.atlassian.net/browse/KIT-127 make it work with connectedCallback
```
Comment (btaillon) [inline packages/atomic/src/components/result-template-components/result-template-decorators.tsx:76]: Typo?
```suggestion
    // TODO: https://coveord.atlassian.net/browse/KIT-1277 make it work with connectedCallback
```

**Review by btaillon — APPROVED** [review]
Nice! I like it

**Pull request merged**